### PR TITLE
Remove impl_Insertable macro

### PR DESCRIPTION
**Pull request by weiznich**

### CODE REVIEW

File: diesel/src/migrations/schema.rs [hunk 1]
```diff
@@ -7,10 +7,53 @@ table! {
...
+use query_builder::insert_statement::{IntoInsertStatement, InsertStatement};
+use expression::helper_types::AsNullableExpr;
+
+impl<'insert, 'a, DB> Insertable<self::__diesel_schema_migrations::table, DB>
+    for &'insert NewMigration<'a>
+    where DB: Backend,
+               (ColumnInsertValue<self::__diesel_schema_migrations::version,
+                   AsNullableExpr<&'insert &'a str,
+                       self::__diesel_schema_migrations::version>>, ): InsertValues<DB>,
+{
```
Comment (killercup) [inline diesel/src/migrations/schema.rs]: This is pretty much unparsable to me. Maybe format it like this?

```rust
type Table = self::__diesel_schema_migrations::table;
type Version = self::__diesel_schema_migrations::version;

impl<'ins, 'a, DB> Insertable<Table, DB> for &'ins NewMigration<'a> where
    DB: Backend,
    (ColumnInsertValue<
        Version,
        AsNullableExpr<&'ins &'a str, version>
    >,): InsertValues<DB>,
```

File: diesel_codegen/src/insertable.rs [hunk 2]
```diff
@@ -15,19 +17,176 @@ pub fn derive_insertable(item: syn::MacroInput) -> quote::Tokens {
...
+
+    match a.field_kind() {
+        "option" => {
+            quote!(
+                match #field_access {
+                    value @ &Some(_) => {
+                        ColumnInsertValue::Expression(
+                            #column,
+                            AsExpression::<<<#column as Expression>::SqlType
+                                as IntoNullable>::Nullable>::as_expression(value)
```
Comment (killercup) [inline diesel_codegen/src/insertable.rs]: Can you try not to break the line here? I'd split this up into

```rust
type NullableColumn = AsExpression::<
    <<#column as Expression>::SqlType as IntoNullable>::Nullable
>;
```

and then use `NullableColumn::as_expression(value)`

File: diesel/src/migrations/schema.rs [hunk 3]
```diff
@@ -7,10 +7,53 @@ table! {
...
+
+    fn values(self) -> Self::Values {
+        use expression::{AsExpression, Expression};
+        use insertable::ColumnInsertValue;
+        use types::IntoNullable;
+        let ref version = self.0;
+        (ColumnInsertValue::Expression(self::__diesel_schema_migrations::version,
+            AsExpression::<<<self::__diesel_schema_migrations::version
+                as Expression>::SqlType as IntoNullable>::Nullable>
+            ::as_expression(version)),)
```
Comment (killercup) [inline diesel/src/migrations/schema.rs]: Can you try not to break this at `as` and `::`? It's really hard to read this way. Using the `Table` alias from abovem you could write

```rust
(ColumnInsertValue::Expression(Version,
    AsExpression::<
        <<Version as Expression>::SqlType as IntoNullable>::Nullable
    >::as_expression(version)
),)
```

File: diesel_codegen/src/insertable.rs [hunk 4]
```diff
@@ -15,19 +17,176 @@ pub fn derive_insertable(item: syn::MacroInput) -> quote::Tokens {
...
+                    &None => ColumnInsertValue::Default(#column),
+                }
+            )
+        }
+        "regular" => {
+            quote!(
+                ColumnInsertValue::Expression(
+                    #column,
+                    AsExpression::<<<#column as Expression>::SqlType
+                        as IntoNullable>::Nullable>::as_expression(#field_access)
```
Comment (killercup) [inline diesel_codegen/src/insertable.rs]: Pls don't break at `as`. See above.

File: diesel_codegen/src/insertable.rs [hunk 5]
```diff
@@ -15,19 +17,176 @@ pub fn derive_insertable(item: syn::MacroInput) -> quote::Tokens {
...
+                         .iter()
+                         .map(|l|{
+                             syn::WherePredicate::RegionPredicate(
+                                 syn::WhereRegionPredicate{
+                                     lifetime: l.lifetime.clone(),
+                                     bounds: vec![insert.lifetime.clone()]
+                                 }
+                             )
+                         }))
+        .build();
```
Comment (killercup) [inline diesel_codegen/src/insertable.rs]: Can you use block ident? I'd write this as

```rust
…
.with_predicates(
    model.generics.lifetimes
    .iter()
    .map(|l|{
        syn::WherePredicate::RegionPredicate(
            syn::WhereRegionPredicate{
                lifetime: l.lifetime.clone(),
                bounds: vec![insert.lifetime.clone()]
            }
        )
    })
)
.build();
```

(Also note space before `{` and trailing comma in `WhereRegionPredicate`)

File: diesel_codegen/src/insertable.rs [hunk 6]
```diff
@@ -15,19 +17,176 @@ pub fn derive_insertable(item: syn::MacroInput) -> quote::Tokens {
...
+            Op,
+             Ret
+        > for &'insert #struct_ty
+        {
+            type InsertStatement =
+                diesel::query_builder::insert_statement::InsertStatement<
+                    #table_name::table,
+                    Self,
+                    Op,
+                    Ret
```
Comment (killercup) [inline diesel_codegen/src/insertable.rs]: Trailing commas are good :)

File: diesel_codegen/src/insertable.rs [hunk 7]
```diff
@@ -15,19 +17,176 @@ pub fn derive_insertable(item: syn::MacroInput) -> quote::Tokens {
...
+                &'insert #field_ty,
+                #table_name::#column_name,
+            >
+        >)
+}
+
+fn make_column_insert_value(
+    a: &Attr,
+    table_name: &syn::Ident,
+    struct_name: &str
```
Comment (killercup) [inline diesel_codegen/src/insertable.rs]: Trailing comma

File: diesel_codegen/tests/insertable.rs [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    users {
+        id -> Integer,
+        name -> VarChar,
+        hair_color -> Nullable<VarChar>,
+    }
+}
+
+#[test]
+fn simple_struct_definition() {
+
```
Comment (killercup) [inline diesel_codegen/tests/insertable.rs]: ✂️

File: diesel/src/migrations/schema.rs (hunk 3 above)
Comment (killercup) [inline diesel/src/migrations/schema.rs]: I'm a big fan of type aliases btw. They can be used to make nice "reverse" associated type accessors like this (IIRC):

```rust
type SqlType<T> = <T as Expression>::SqlType;
type Nullable<T> = <T as IntoNullable>::Nullable;
```

**Comment by weiznich** [comment]
Any news on this?

**Comment by sgrif** [comment]
I'd like to postpone these for the time being. I think there are two steps I'd like to take as separate PRs before we take this step:

- Move the tests for the bang macro to instead use the `#[derive]` (and presumably move them to the `diesel_codegen` crate)
- Remove the struct parsing code from diesel and only leave the entry point that codegen uses.

That would make our job of reviewing significantly easier, help point out gaps in our test coverage, and make me more comfortable with this step.

**Pull request closed without merging**